lambdas/pkgpush/index.py

```diff
@@ -172,6 +179,51 @@
 logger = get_quilt_logger()
 
 
+def calculate_pkg_hashes(s3_client, pkg):
+    entries = []
+    for lk, entry in pkg.walk():
+        if entry.hash is not None:
+            continue
+        if entry.size > S3_HASH_LAMBDA_MAX_FILE_SIZE:
+            raise FileTooLargeForHashing(lk)
+
+        entries.append(entry)
+
+    with concurrent.futures.ThreadPoolExecutor(max_workers=S3_HASH_LAMBDA_CONCURRENCY) as pool:
+        fs = [
+            pool.submit(calculate_pkg_entry_hash, s3_client, entry)
+            for lk, entry in pkg.walk()
+            if entry.hash is None
+        ]
+        for f in concurrent.futures.as_completed(fs):
+            f.result()
+
+
+def calculate_pkg_entry_hash(s3_client, pkg_entry):
+    pk = pkg_entry.physical_key
+    params = {
+        'Bucket': pk.bucket,
+        'Key': pk.path,
+    }
+    if pk.version_id is not None:
+        params['VersionId'] = pk.version_id
+    url = s3_client.generate_presigned_url(
+        ClientMethod='get_object',
+        ExpiresIn=S3_HASH_LAMBDA_SIGNED_URL_EXPIRES_IN,
+        Params=params,
+    )
+    pkg_entry.hash = {
+        'type': 'SHA256',
+        'value': invoke_hash_lambda(url),
+    }
+
+
+def invoke_hash_lambda(url):
+    resp = lambda_.invoke(FunctionName=S3_HASH_LAMBDA, Payload=json.dumps(url))
+    assert 'FunctionError' not in resp
+    return json.load(resp['Payload'])
```

Comment:
i think we can tighten this up a bit. first the S3_HASH_LAMBDA should take a hash type ("SHA256" for now) in case we change that. and that brings me to my second point. it's somewhat opaque to return a string fro Payload and just assume it's a SHA256.

so this function should do two things differently:
1. Return a full hash value: `{'type': 'SHA256', 'value': 'somethingsomethingsomething'}`
2. Verify that the hash is the right length and composition relative to the type (JSON schema good for this; we already use it to validate query params)

in this way invoke_hash_lambda will have a clearer contract and better ownership of the inputs and outputs.

Reply:
1. I think making hash lambda take/return hash type is overengineering, since I don't expect this function will work with anything besides sha256 anytime soon (actually I think it will never be needed)
2. if we need a sort of sanity checks the right place for them seems quilt3 API

Reply:
well returning a full hash object is a quick change and it's more modular. like even if we don't pass SHA256 in, we should return it `type` so that the lambda is clear about what it's returning.

i still think it's important to check `len(hash.value)` at the least.

if for some reason there is a bug here we will be making make manifests that are impossible to reproduce b/c hash is unattainable.

Reply:
> well returning a full hash object is a quick change and it's more modular. like even if we don't pass SHA256 in, we should return type so that the lambda is clear about what it's doing.

Lambda gets URL and returns sha256, that's its API. It knows nothing about `PackageEntry.hash`. `invoke_hash_lambda()` is a simple wrapper for hash lambda. I could replace `S3_HASH` with `S3_SHA256` if it would make someone's life easier.

> i don't really understand why this couldn't be SHA-3 at some point.

It could be but I don't think it would be anytime soon so I don't see a point in this generalization for hash type.

Reply:
> i still think it's important to check len(hash.value) at the least.

Still not understand why it should be here and not in quilt3 API.

Reply:
Because defensive programming. In many cases (e.g. pure catalog) `quilt3` is not even in the picture.